reactor-core/src/main/java/reactor/core/publisher/FluxSwitchOnFirst.java

```diff
@@ -523,58 +559,141 @@ public void onError(Throwable throwable) {
                 parent.cancel();
             }
 
-            inner.onError(throwable);
+            delegate.onError(throwable);
         }
 
         @Override
         public void onComplete() {
-            if (!parent.done) {
+            if (!parent.done && cancelSourceOnComplete) {
                 parent.cancel();
             }
 
-            inner.onComplete();
+            delegate.onComplete();
+        }
+
+        @Override
+        public void request(long n) {
+            long r = this.requested;
+
+            if (r > STATE_REQUESTED) {
+                long u;
+                for (;;) {
+                    if (r == Long.MAX_VALUE) {
+                        return;
+                    }
+                    u = Operators.addCap(r, n);
+                    if (REQUESTED.compareAndSet(this, r, u)) {
+                        return;
```

Comment:
shouldn't it break from the `for(;;)` loop rather than completely exiting?
I get the `return` in the case where we detect we're running in unbounded mode, but here we're aggregating requests and it is not clear to me how they will be propagated upstream...

Reply:
if it is a long max value then it means that it was requested and not consumed yet. Therefore we have simply returned without checking anything. Actually, I can add tests which check the correctness just in case you are worried about that path

Reply:
what guarantees that the `n` you just added to `r` in `addCap` are actually propagated upstream? since you're returning, and `r > STATE_REQUESTED`, there's no chance the `tryRequest` will swap the new `r` for `-1` and perform `s.request(r)`, so that amount `n` is lost. Am I missing something?

Reply:
I'm not sure. 

Let's consider the following:

State 1: No subscription actual make a request on the given subscription so we in into the request method:

[Line 576](https://github.com/reactor/reactor-core/pull/2019/files/877b27053c1efc5fd287d1617535114e32ccb1d1#diff-c590f0f2fbe1d419bdccc521a1894278R576) : we read the current request size

[Line 578](https://github.com/reactor/reactor-core/pull/2019/files/877b27053c1efc5fd287d1617535114e32ccb1d1#diff-c590f0f2fbe1d419bdccc521a1894278R576) we check that r is higher than -1 and in into for-loop

Cases: [Line 585](https://github.com/reactor/reactor-core/pull/2019/files/877b27053c1efc5fd287d1617535114e32ccb1d1#diff-c590f0f2fbe1d419bdccc521a1894278R576): perform CAS operation which means if it is (1 - true, 2 - false): 

 1.  then we have won racing and the last observed request was higher than -1 so there is no upstream subscription and we can exit request method 
 2.  then we read the request field at [Line 589](https://github.com/reactor/reactor-core/pull/2019/files/877b27053c1efc5fd287d1617535114e32ccb1d1#diff-c590f0f2fbe1d419bdccc521a1894278R589) and perform a check on request is less than 0 (which means it is either canceled or we lost the racing so we have to break the loop and go straight to reading subscription `s` and perform a direct `request(n)`.

Reply:
thanks for the explanations, comments and face-to-face review, this is clear to me now 👍